### PR TITLE
[Fixes #8564] `Metrics/AbcSize`: Add optional discount for repeated "attributes"

### LINKED ISSUE #8564: Redefine what counts as a branch on ABC Size
I was thinking if it would be possible to not count getter methods as a branch. I know this has been asked before (can't find the issue now, but I de remember reading something about it) and I'm just wondering if it may be possible with a plugin instead of a configuration setting.

The reason for it is pretty simple, take this method for example

```ruby
def search
  @posts = model.active.visible_by(current_user)
    .search(params[:q], skip: @skip)

  @posts = model.some_process(@posts, current_user)
  @posts = model.another_process(@posts, current_user)

  render 'pages/search/page', locals: { model: model }
end
```

It's ABC would be `<3, 15, 0> = 15.3`

Now, if instead of calling model and current_user every time, you assign their values to a variable you get:

```ruby
def search
  the_model = model
  user = current_user
  @posts = the_model.active.visible_by(user)
    .search(params[:q], skip: @skip)

  @posts = the_model.some_process(@posts, user)
  @posts = the_model.another_process(@posts, user)

  render 'pages/search/page', locals: { model: the_model }
end
```

It's ABC would be `<5, 10, 0> = 11.18`

So the reason this bothers me is that the last method is considered the least complex one, but in reality it takes more time for someone to read through it.

Branches are defined as "an explicit forward program branch out of scope", I would say getter methods in ruby are not really "explicit" as they pretty much look like just any variable. But I'm aware that's highly a matter of personal opinion.

So wrapping up, is there a way to not count any getter method (defined for this purpose as any method that receives no arguments) as branch while calculating ABC? I had a look at `AbcSizeCalculator` class but I can't figure out where the `node.type` comes from in the `branch?(node)` method

### PR DESCRIPTION
**Pull request by marcandre**
That was more involved that I had hoped for...

I'm unsure as to the naming of the new config option; within the code I feel that `discount_repeated_attributes` was easiest to think about, but for the config I'm proposing `CountRepeatedAttributes` instead of `Discount...`. Maybe `IgnoreRepeatedAttributes` would be better?

### CODE REVIEW

**Comment by marcandre** [comment]
Also wondering if certain methods should never be considered as attributes, say `pop shift drop receive new save sort delete next reset flush`. Most are unlikely to be repeated on the same object, but some could:

```ruby
def process
  foo(queue.pop)
  bar(queue.pop) # This second `pop` should always be counted...
  # ...
end
```

Worth it?

File: lib/rubocop/cop/metrics/abc_size.rb [hunk 1]
```diff
@@ -7,6 +7,27 @@ module Metrics
       # configured maximum. The ABC size is based on assignments, branches
       # (method calls), and conditions. See http://c2.com/cgi/wiki?AbcMetric
       # and https://en.wikipedia.org/wiki/ABC_Software_Metric.
+      #
+      # You can have repeated "attributes" calls count as a single "branch".
+      # For this purpose, attributes are any method with no argument; no attempt
```
Comment (bbatsov) [inline lib/rubocop/cop/metrics/abc_size.rb:12]: and an implicit/current object receiver I guess.

**Review by bbatsov — APPROVED** [review]
I like your proposal. Definitely makes a lot of sense.

File: lib/rubocop/cop/metrics/abc_size.rb (hunk 1 above)
Comment (marcandre) [reply lib/rubocop/cop/metrics/abc_size.rb:12]: Not necessarily:
```ruby
foo.bar.baz # => 3 branches
foo.bar.qux # => 1 more branch only (`foo` and `bar` are "repeated")

n = 42
n.to_s # => 1 branch
n.to_s.to_i # => just 1 more (`to_s` "repeated")
```

**Comment by marcandre** [comment]
> I like your proposal. Definitely makes a lot of sense.

Thanks. So `CountRepeatedAttributes` is good?

**Comment by bbatsov** [comment]
I think it's fine - I don't have a strong preference for it or `IgnoreRepeatedAttributes`.

**Comment by marcandre** [comment]
Last remaining question:

> Also wondering if certain methods should never be considered as attributes, say `pop shift drop receive new save sort delete next reset flush`. Most are unlikely to be repeated on the same object, but some could:
> 
> ```ruby
> def process
>   foo(queue.pop)
>   bar(queue.pop) # This second `pop` should always be counted...
>   # ...
> end
> ```
> 
> Worth it?

**Comment by bbatsov** [comment]
Seems like an overkill to me, but I'd be fine with adding some special handling for them (ideally in the form of configuration).

**Comment by marcandre** [comment]
Ok, then this is mergeable as is then 👍

Rebased just to be sure (maybe there are new style issues)

**Comment by bbatsov** [comment]
Great! It's heading to production later today or tomorrow with the next release.

**Pull request merged**